core/src/services/foundationdb/backend.rs

```diff
@@ -0,0 +1,238 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::cmp::min;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use foundationdb::api::NetworkAutoStop;
+use foundationdb::tuple::Subspace;
+use foundationdb::{Database, FdbError, RangeOption};
+use lazy_static::lazy_static;
+
+use futures::StreamExt;
+use std::iter::Iterator;
+
+use crate::raw::adapters::kv;
+use crate::Builder;
+use crate::Error;
+use crate::ErrorKind;
+use crate::Scheme;
+use crate::*;
+
+lazy_static! {
+    static ref FOUNDATIONDB_API_INITIALIZE_HANDLE: NetworkAutoStop =
+        unsafe { foundationdb::boot() };
+}
+
+/// Foundationdb service support.
+#[doc = include_str!("docs.md")]
+#[derive(Default)]
+pub struct FoundationdbBuilder {
+    root: Option<String>,
+}
+
+impl FoundationdbBuilder {
+    /// Set the root for Foundationdb.
+    pub fn root(&mut self, path: &str) -> &mut Self {
+        self.root = Some(path.into());
+        self
+    }
+}
+
+impl Builder for FoundationdbBuilder {
+    const SCHEME: Scheme = Scheme::Foundationdb;
+    type Accessor = FoundationdbBackend;
+
+    fn from_map(map: HashMap<String, String>) -> Self {
+        let mut builder = FoundationdbBuilder::default();
+
+        map.get("root").map(|v| builder.root(v));
+
+        builder
+    }
+
+    fn build(&mut self) -> Result<Self::Accessor> {
+        lazy_static::initialize(&FOUNDATIONDB_API_INITIALIZE_HANDLE);
+
+        let db = foundationdb::Database::default().map_err(|e| {
+            Error::new(ErrorKind::ConfigInvalid, "open foundation db")
+                .with_context("service", Scheme::Foundationdb)
+                .set_source(e)
+        })?;
+
+        let db = Arc::new(db);
+
+        Ok(FoundationdbBackend::new(Adapter { db })
+            .with_root(self.root.as_deref().unwrap_or_default()))
+    }
+}
+
+/// Backend for Foundationdb services.
+pub type FoundationdbBackend = kv::Backend<Adapter>;
+
+#[derive(Clone)]
+pub struct Adapter {
+    db: Arc<Database>,
+}
+
+impl Debug for Adapter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Adapter");
+        ds.finish()
+    }
+}
+
+// Foundationdb has a limit of 100,000 bytes per key
+// and 10,000,000 bytes of affected data each transaction,
+// so we need to split
```

Comment:
Please return an error instead. OpenDAL can't and shouldn't do split work as service side.

Reply:
I can update the code to let OpenDAL store the value as it is. But in this way the normal write test will fail because it will produce larger file. And there's no workaround for foundationdb to increase the limit, such as setting env variable `MEMCACHED_MAX_ITEM_SIZE` for memcached.

I am planning to make the `gen_bytes()` in [utils.rs](https://github.com/apache/incubator-opendal/blob/main/core/tests/behavior/utils.rs#L96) byte size configurable. Specifically, I'd like to define an env variable `OPENDAL_TEST_MAX_ITEM_SIZE` and replace the line:
```rust
// let size = rng.gen_range(1..4 * 1024 * 1024);
let size = rng.gen_range(1..env::var("OPENDAL_TEST_MAX_ITEM_SIZE").unwrap_or(4 * 1024 * 1024));
```

Is this ok?

Reply:
The problem we encountered is that we need to include the maximum size in the capacity so that users are aware of it before writing.

Reply:
Updated code to avoid extra logic to process data, and mentioned in the `Capabilities` section of `docs.md` to mention the limitations.
Also updated `service_test_foundationdb.yml` to turn off the test for now.

Reply:
Thanks a lot!